pkg/nats/stream/producer_client.go

```diff
@@ -0,0 +1,223 @@
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctxWithCancel.Done():
+			log.Ctx(ctxWithCancel).Debug().Msg("Heart beat for NATS based streaming producer client cancelled.")
+			return
+		case <-ticker.C:
+
+			nonActiveStreamIds := make(map[string][]string)
+
+			for c, v := range pc.activeConnHeartBeatRequestSubjects {
+				// Create an empty slice for activeStreamIds
+				activeStreamIds := make(map[string][]string)
+
+				if streamInfoMap, ok := pc.activeStreamInfo[c]; ok {
+					for _, streamInfo := range streamInfoMap {
+						activeStreamIds[streamInfo.RequestSub] = append(activeStreamIds[streamInfo.RequestSub], streamInfo.ID)
+					}
+				}
+
+				heartBeatRequest := HeartBeatRequest{
+					ActiveStreamIds: activeStreamIds,
+				}
+
+				data, err := json.Marshal(heartBeatRequest)
+				if err != nil {
+					log.Ctx(ctx).Err(err).Msg("error while parsing heart beat request in NATS streaming producer client")
+					continue
+				}
+
+				msg, err := pc.Conn.Request(v, data, pc.config.HeartBeatRequestTimeout)
+				if err != nil {
+					log.Ctx(ctx).Err(err).Msg("error while sending heart beat request from NATS streaming producer client")
+					nonActiveStreamIds[c] = []string{}
+					continue
+				}
+
+				var heartBeatResponse ConsumerHeartBeatResponse
+				err = json.Unmarshal(msg.Data, &heartBeatResponse)
+				if err != nil {
+					log.Ctx(ctx).Err(err).Msg("error while  parsing heart beat response from NATS streaming consumer client")
+					continue
+				}
+
+				nonActiveStreamIds[c] = getStringList(heartBeatResponse.NonActiveStreamIds)
+			}
+
+			pc.updateActiveStreamInfo(nonActiveStreamIds)
+		}
+	}
+}
+
+func (pc *ProducerClient) WriteResponse(
+	consumerID string,
+	streamID string,
+	obj interface{},
+	writer *Writer,
+) (int, error) {
+	pc.mu.Lock()
+	streamIds, active := pc.activeStreamInfo[consumerID]
+
+	if !active {
+		return 0, fmt.Errorf("no stream ids exist to write for consumerId=%s", consumerID)
+	}
+	pc.mu.Unlock()
+
+	_, active = streamIds[streamID]
+	if !active {
+		return 0, fmt.Errorf("streamId %s is now closed", streamID)
+	}
+
+	return writer.WriteObject(obj)
+}
+
+func (pc *ProducerClient) updateActiveStreamInfo(nonActiveStreamIds map[string][]string) {
+	// Create a map to store the streams that need to be deleted.
+	streamsToDelete := make(map[string][]string)
+
+	for connID, nonActiveIdsList := range nonActiveStreamIds {
+		// Create a map for quick lookup of non-active streams.
+		nonActiveMap := make(map[string]bool)
+		for _, id := range nonActiveIdsList {
+			nonActiveMap[id] = true
+		}
+
+		if streamInfo, ok := pc.activeStreamInfo[connID]; ok {
+			for streamID := range streamInfo {
+				// If the stream is not active, store it for deletion.
+				if nonActiveMap[streamID] {
+					streamsToDelete[connID] = append(streamsToDelete[connID], streamID)
+				}
+			}
+		}
+	}
+
+	// Delete all inactive streams with minimal lock contention.
+	pc.mu.Lock()
```

Comment:
Since we are doing the locking at very tail end of the function, we may be fine here.

Mainly because  `defer` places the function call on a stack to be executed when the surrounding function finishes.

But I agree we can run into a situation where if delete panics, we won't be unlocking.

Reply:
Will fix it.